src/v/raft/tests/raft_fixture.h

```diff
@@ -470,30 +467,26 @@ class raft_fixture
         };
         return ss::do_with(
           retry_state{},
-          [this, deadline, f = std::forward<Func>(f), backoff](
-            retry_state& state) mutable {
+          std::forward<Func>(f),
```

Comment:
nit: I wonder if this should be a forward or a move

Reply:
I think it should be a forward. The lazy monkey argument would be that all subsequent calls forward it, so why should we break this chain? Really, they seem to do it for a reason, the reason being `do_with` actually working "well" with lvalue references. It accepts an lvalue reference and never copies or moves the object. In this example the object is not copiable, and after calling with an lvalue reference we can check it has not been moved-from either:
```
struct kv1_stm_fixture : kv_stm_fixture<kv1_stm_t> {
    ...
    ss::future<int> foo_rvref() {
        std::unique_ptr<int> uptr = std::make_unique<int>(5);
        return ss::do_with(std::move(uptr), [](const std::unique_ptr<int>& p) {
            return foo_impl(p);
        });
    }

     // the test fails with `AddressSanitizer: stack-use-after-return` if we declare it in the function body
    std::unique_ptr<int> member_uptr = std::make_unique<int>(5);
    ss::future<int> foo_lvref() {
        return ss::do_with(
          std::ref(member_uptr),
          [](const std::unique_ptr<int>& p) { return foo_impl(p); });
    }

    static ss::future<int>
    foo_impl(const std::unique_ptr<int>& non_copiable_parameter) {
        co_await ss::sleep(1s);
        vlog(kvlog.info, "lv");
        co_return *non_copiable_parameter.get();
    }

    static ss::future<int>
    foo_impl(std::unique_ptr<int>&& non_copiable_parameter) {
        co_await ss::sleep(1s);
        vlog(kvlog.info, "rv");
        co_return *std::move(non_copiable_parameter).release();
    }
};

TEST_F_CORO(kv1_stm_fixture, test_mux_state_machine_simple_scenarios) {
    ASSERT_EQ_CORO(co_await foo_lvref(), 5);
    ASSERT_TRUE_CORO(member_uptr);
    ASSERT_EQ_CORO(co_await foo_rvref(), 5);
...
}
```
I put "well" in quotes because, when called with an lvalue reference, `do_with` does not attempt to extend the object lifetime (and it's not that it has a chance to). So, despite `do_with` it remains the caller's responsibility to ensure the object is alive until the future completion.

I find it handy, as it allows us code universal functions like `retry_with_leader` to accept lvalue references to long-life objects as well as rvalue references to temporaries. We don't have to code separate versions of such functions, one with `do_with` and another one without, we just need to forward in there.

The only problem is that this behaviour of `do_with` is undocumented.

Reply:
> I put "well" in quotes because, when called with an lvalue reference, do_with does not attempt to extend the object lifetime (and it's not that it has a chance to). So, despite do_with it remains the caller's responsibility to ensure the object is alive until the future completion.

Right.. I was wondering whether we should disallow lvalue refs here to avoid this footgun. To your point that it is handy to just forward and let the callers handle the lifetime seems like a fair argument, especially in this test code.

Reply:
TBH I only checked forwarding/moving of arguments, not the function itself. But I think there's no harm in making them behave consistently as, eventually they anyway form a closure together.